recitation_6/file-io.md

```diff
@@ -0,0 +1,234 @@
+
+The 3 basic channels (often called "streams") as well as their corresponding integer representations are:
+
+-   **(0) stdin (standard input)**  This stream is for incoming data, which often comes from the keyboard but can also be from other sources (e.g. through piping & redirection)
+-   **(1) stdout (standard output)**  This stream is for outgoing data, and normally goes to the terminal screen but does not necessarily have to. This stream is buffered which means it is not sent to the terminal until a new line character is sent. This means if you use  `printf("hello")`  you likely will not see it until the end of your program is reached.
+-   **(2) stderr (standard error)**  This stream is for error messages and is not buffered, meaning any characters written to it will immediately be flushed to their destination. This destination is often the terminal screen, but can be other locations as well.
+
+If you wish to interact with these buffers you will need to  `#include <stdio.h>`. This library defines standard functions such as  `printf()`  `scanf()`  and others which you may or may not have already used.
+
+### Redirection and Piping
+
+Lots of input and output to/from programs is visible in the shell (your terminal screen). That being said, it's possible to redirect the source of stdin or the destination of stderr and stdout.
+
+There are several kinds of redirection possible from the console:
+
+-   `>`  redirects the standard output of the left argument to the right argument. For example  `echo "hello world" > hi.txt`  will write the words 'hello world' to a file called  `hi.txt`, and 'hello world' will not appear on the terminal screen. 
+-   `2>`  is similar, except that it takes the standard error and redirects it, rather than the standard output.
+-   `<`  does 'input redirection', meaning that the contents of the file on the right of the operator serves as the standard input of the left argument. For example, after running the example above,  `cat < hi.txt`  will print "hello world" to the console.
+-   `2>&1`  will redirect stderr to the same location as stdout.
+-   `>>`  will append the output to a file instead of overwriting the file.
+-  `|`  will feed the output from the program on the left as input to the program on the right. For example the `head -n` command will print as output the first n number of data of its input. Running `ls | head -3` will print up to 3 files in your working directory. 
+
+### Formatting
+
+printf and scanf both use format strings to specify what how to format their output. They also both accept variable arguments. All arguments to scanf  **must**  be pointers whereas arguments to printf should be values (in the case of numbers) or  `char *`  in the case of strings. Pages 153-154 in the K&R explain how to format your format strings for  `printf()`  and 157-158 explain formatting for  `scanf()`. Make sure you can identify the following two format strings:
```

Comment:
"whereas arguments to printf should be values (in the case of numbers) or `char *` in the case of strings."
This is not true, printf("%p", &x), &x could be an int * or whatever pointer for that matter.